docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,340 @@
+    @gtscript.stencil(backend=backend_type)​
+    def main_ub(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ub: FIELD_TYPE, dt5: float):​
+        with computation(PARALLEL), interval(...):​
+            ub = dt5 * (uc[0, -1, 0] + uc - ​(vc[-1, 0, 0] + vc) * cosa) * rsina​
+
+    @gtscript.stencil(backend=backend_type)​
+    def x_edge_ub(ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float)::
+        with computation(PARALLEL), interval(...):​
+            ub = dt5 * (ut[0, -1, 0] + ut)​
+
+    @gtscript.stencil(backend=backend_type)​
+    def y_edge_ub(ut: FIELD_TYPE, ub: FIELD_TYPE, *, dt4: float):
+        with computation(PARALLEL), interval(...):
+            ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+    # Requires new python function to call stencils on each region
+    def ubke(uc, vc, ut, ub, dt5, dt4, grid):​
+        domain_y_edge = (grid.ni, 1, grid.nz)
+        domain_x_edge = (1, grid.nj, grid.nz)
+        main_ub(uc, vc, grid.cosa, grid.rsina, ub, dt5=dt5, ​
+                origin=(grid.local_istart, grid.local_jstart, 0),
+                domain=(grid.ni, grid.nj, grid.nz))​
+        if grid.west_edge:​
+            x_edge_ub(ut, ub, dt5=dt5, ​origin=(grid.local_istart, grid.local_jstart, 0), ​domain=domain_x_edge)​
+        if grid.south_edge:
+            y_edge(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jstart, 0), domain=domain_y_edge)
+        if grid.north_edge:
+            y_edge_ub(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jend, 0), domain=domain_y_edge)
+        if grid.east_edge:
+            x_edge_ub(ut, ub, dt5=dt5, origin=(grid.local_iend, grid.local_jstart, 0), domain=domain_x_edge)
+
+This example motivates a number of requirements for the feature. It must be able to
+
+1. specify regions in the context of a global iteration space (for global boundary conditions) using different `with` blocks inside a single stencil
+2. specify cleanly regions at and near the edge of compute domains
+3. read from and write to fields inside halo regions
+
+The specific feature that we are proposing is the addition of a `region()` specifier that complements the `interval()`, but treats the horizontal instead of the vertical.
+Arguments passed to the `region()` specifier are Python expressions resolved at compile time, which result in objects specifying the local index region for the contained computation.
+In order to make this friendlier for applications, they can provide methods or classes that provide these objects specifying the local index region.
+
+Using this concept, the example above is transformed into:
+
+.. code-block:: python
+
+    @gtscript.stencil(backend=backend_type) # Single stencil
+    def ubke(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float, dt4: float):
+        with computation(PARALLEL):
+            with interval(...): # applies to the compute domain as it does now
```

Comment:
I miss some semantics here. 

- The `with` clause without the `region` is executed in all points **except** the ones indicated by ones indicating a `region`?
- It looks to me that every time you have, like here, two regions on two different axes you end up with a non-empty intersection. In this case the assumption of the execution order, mentioned later, should become part of the specification (unless I'm missing something here): either the first is the one executed or the second (it not clear to me why the non-empty intersection is treated separately later on), or the last one is. In the latter case, the user can _easily_ introduce a third `region` with the corner and they would now that that would be applied.
- Additionally, an intersection is always present when a `region` and an `interval` is specified. What is the semantic here? Who has the precedence? In other words: how is the address space partitioned? May a  picture would be useful to understand. In general, even assuming to just do boundaries all around (horizontal and vertical) you end up with 27 regions (assuming the _main_ one does not have a region specifier, they are 26), with only 3, or 6, region and interval specifiers, how do you sort out how deal with all of them?
- I see that intersections need to be taken care of, since a user can specify multiple times a same region with different computations, but still, the problem of the _natural_ intersections arising just because the horizontal and vertical regions intersects is not clear to me.